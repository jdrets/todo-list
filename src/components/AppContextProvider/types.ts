import React from 'react'

export interface TasksTypes {
  status: string
  priority: string
  dueDate?: string
  title: string
  description: string
}

export interface AppContextProviderTypes {
  tasks: TasksTypes[]
  setTasks: React.SetStateAction<any>
  newTaskModal: boolean
  setNewTaskModal: React.SetStateAction<any>
}
