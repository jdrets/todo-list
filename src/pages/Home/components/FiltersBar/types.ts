import { SetStateAction } from 'react'

export interface FiltersTypes {
  label: string
  onClose: () => void
  // TODO: add filters types
}

export interface FiltersBarTypes {
  filters?: any
  setFilters: SetStateAction<any>
  openFilterModal: () => void
}
