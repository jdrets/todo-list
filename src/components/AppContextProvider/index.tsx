import React, { FunctionComponent, useState, createContext, useEffect } from 'react'
import _ from 'lodash'

import TaskService from '../../services/taskService'
import { AppContextProviderTypes } from './types'

export const PageContext = createContext({} as AppContextProviderTypes)

const AppContextProvider: FunctionComponent<any> = ({ children, value = {} }) => {
  const [tasks, setTasks] = useState([])
  const [showCreateTaskModal, setShowCreateTaskModal] = useState(false)

  useEffect(() => {
    const storage = TaskService.get()
    setTasks(storage)
  }, [])

  const updateContextWithNewTask = (task) => {
    const newTasks = _.clone(tasks)
    newTasks.push(task)
    setTasks(_.reverse(newTasks))
  }

  const contextValues = {
    tasks,
    showCreateTaskModal,
    updateContextWithNewTask,
    setShowCreateTaskModal,
    ...value
  }

  return (
    <PageContext.Provider value={contextValues}>{children}</PageContext.Provider>
  )
}

export default AppContextProvider
