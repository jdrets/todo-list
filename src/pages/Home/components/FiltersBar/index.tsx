import React, { FunctionComponent, useEffect, useState } from 'react'
import _ from 'lodash'
import { Filter as FilterIcon } from 'react-ionicons'
import uuid from 'react-uuid'

import { FiltersBarTypes } from './types'
import {
  Wrapper,
  FilterButton,
  FilterButtonLabel
} from './styles'

import Chip from '../../../../components/Chip'

const FiltersBar: FunctionComponent<FiltersBarTypes> = ({
  filters,
  setFilters,
  openFilterModal
}) => {
  const [chips, setChips] = useState([])

  useEffect(() => {
    setChips(_.toPairs(filters))
  }, [filters])

  const deleteFilter = (key) => {
    setFilters({
      ...filters,
      [key]: null
    })
  }

  return (
    <Wrapper>
      <section>
        {
          chips.map(filter => (
            <Chip
              key={uuid()}
              deleteFilter={deleteFilter}
              filter={filter}
            />
          ))
        }
      </section>
      <FilterButton onClick={openFilterModal}>
        <FilterIcon />
        <FilterButtonLabel>Filtros</FilterButtonLabel>
      </FilterButton>
    </Wrapper>
  )
}

export default FiltersBar
