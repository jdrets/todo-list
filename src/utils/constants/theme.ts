const palette = {
  purpleDark: '#4B00BC',
  purple: '#6600FF',
  purpleMedium: '#8533FF',
  purpleLight: '#E3D0FF',
  green: '#59BC6F',
  greenLight: '#E6F5E2',
  red: '#EB5757',
  redLight: '#FCE4E4',
  yellow: '#DD9900',
  yellowLight: '#FEF1D8',
  white: '#FFFFFF',
  black: '#000000',
  gray: '#898989',
  grayLight: '#E0E0E0'
}

const theme = {
  colors: {
    primaryColor: palette.purple,
    primaryColorMedium: palette.purpleMedium,
    primaryColorLight: palette.purpleLight,
    primaryColorDark: palette.purpleDark,
    ...palette
  },
  layout: {
    breakpoints: {
      medium: '768px'
    },
    mobile: {
      width: '100%',
      padding: '0 24px'
    },
    desktop: {
      width: '1050px'
    }
  }
}

export default theme
