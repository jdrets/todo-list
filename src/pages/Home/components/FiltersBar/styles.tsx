import styled from 'styled-components'

import theme from '../../../../utils/constants/theme'

export const Wrapper = styled.section`
  align-items: center;
  display: flex;
  justify-content: space-between;
  padding: 6px 0 12px;
  position: relative;
  z-index: 2;
`

export const FilterButton = styled.button`
  align-items: center;
  background-color: transparent;
  border: none;
  cursor: pointer;
  display: flex;
  outline: none;
  
  svg {
    align-items: center;
    display: flex;
  }
`

export const FilterButtonLabel = styled.span`
  display: none;
  
  @media (min-width: ${theme.layout.breakpoints.medium}) {
    display: block;
    font-weight: 500;
    margin-left: 6px;
  }
`
