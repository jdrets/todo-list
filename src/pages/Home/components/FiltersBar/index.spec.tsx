import React from 'react'
import { ThemeProvider } from 'styled-components'

import { fireEvent, render } from '@testing-library/react'

import FiltersBar from '.'
import theme from '../../../../utils/constants/theme'
import AppContextProvider from '../../../../components/AppContextProvider'

const setup = (mockData = {}) => {
  const mock = {
    openFilterModal: jest.fn(),
    setFilters: jest.fn(),
    filters: {
      title: 'dummy title'
    }
  }
  return render(
    <ThemeProvider theme={theme}>
      <AppContextProvider>
        <FiltersBar {...mock} {...mockData} />
      </AppContextProvider>
    </ThemeProvider>
  )
}

describe('<FiltersBar />', () => {
  test('should match with snapshot', () => {
    const { container } = setup()

    expect(container).toMatchSnapshot()
  })

  test('should call setFilters when delete button is clicked', () => {
    const setFiltersMock = jest.fn()
    const { getByTestId } = setup({
      setFilters: setFiltersMock
    })

    const deleteButton = getByTestId('chip-close-button')

    fireEvent.click(deleteButton)

    expect(setFiltersMock).toHaveBeenCalled()
  })
})
