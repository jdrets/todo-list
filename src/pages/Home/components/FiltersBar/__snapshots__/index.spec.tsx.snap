// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FiltersBar /> should match with snapshot 1`] = `
<div>
  <section
    class="sc-gtssRu gPLxaA"
  >
    <section>
      <div
        class="sc-eCApGN jEElRs"
      >
        <div
          class="sc-jSFkmK itaXqv"
        >
          <div
            class="sc-gKAblj ebCotd"
          >
            title
            : 
            dummy title
          </div>
          <div
            class="sc-iCoHVE gSOozR"
            data-testid="chip-close-button"
          >
            <span
              class="sc-bdnylx kpuqbH"
              color="#000000"
              height="22px"
              rotate="0"
              width="22px"
            >
              <svg
                class=""
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                
                <path
                  d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
    </section>
    <button
      class="sc-dlnjPT eLxryg"
      data-testid="open-filters-modal-button"
    >
      <span
        class="sc-bdnylx kpuqbH"
        color="#000000"
        height="22px"
        rotate="0"
        width="22px"
      >
        <svg
          class=""
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <path
            d="M472 168H40a24 24 0 010-48h432a24 24 0 010 48zM392 280H120a24 24 0 010-48h272a24 24 0 010 48zM296 392h-80a24 24 0 010-48h80a24 24 0 010 48z"
          />
        </svg>
      </span>
      <span
        class="sc-hKFyIo jMlnGb"
      >
        Filters
      </span>
    </button>
  </section>
</div>
`;
